crates/ruff/src/args.rs

```diff
@@ -594,12 +680,80 @@ fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
     }
 }
 
+/// --config arguments passed via the CLI
+/// Users may pass 0 or 1 paths to a configuration file,
+/// and an arbitrary number of "inline TOML" options
+///
+/// For example:
+///
+/// ```sh
+/// ruff check --config "path/to/pyproject.toml" --config "max-line-length=90" --config "isort.case-sensitive=false"
+/// ```
+#[derive(Clone, Debug)]
+pub enum ConfigOption {
+    PathToConfigFile(PathBuf),
+    ConfigOverride(Box<Options>),
+}
+
+#[derive(Clone)]
+pub struct ConfigOptionParser;
+
+impl ValueParserFactory for ConfigOption {
+    type Parser = ConfigOptionParser;
+
+    fn value_parser() -> Self::Parser {
+        ConfigOptionParser
+    }
+}
+
+impl TypedValueParser for ConfigOptionParser {
+    type Value = ConfigOption;
+
+    fn parse_ref(
+        &self,
+        cmd: &clap::Command,
+        arg: Option<&clap::Arg>,
+        value: &std::ffi::OsStr,
+    ) -> Result<Self::Value, clap::Error> {
+        let value = value
+            .to_str()
+            .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;
+        let path_to_config_file = PathBuf::from(value);
+        if path_to_config_file.exists() {
+            return Ok(ConfigOption::PathToConfigFile(path_to_config_file));
+        }
+        if let Ok(option) = toml::from_str(value) {
+            Ok(ConfigOption::ConfigOverride(option))
+        } else {
+            let mut new_error =
+                clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
+            if let Some(arg) = arg {
+                new_error.insert(
+                    clap::error::ContextKind::InvalidArg,
+                    clap::error::ContextValue::String(arg.to_string()),
+                );
+            }
+            new_error.insert(
+                clap::error::ContextKind::InvalidValue,
+                clap::error::ContextValue::String(value.to_string()),
+            );
+            new_error.insert(
+                clap::error::ContextKind::Suggested,
+                clap::error::ContextValue::StyledStrs(vec![
+                    "The `--config` flag must either be a path to a `.toml` configuration file or a TOML string providing configuration overrides".into(),
+                    format!("The path `{value}` does not exist on your filesystem").into()
```

Comment:
I'm conflicted if we should use the TOML terminology here or should instead say that it needs to be a config option key `=` value pairs. I at least wouldn't immediately understand what a TOML string is (`"abcd"?`)